### PR TITLE
Persist denotations.

### PR DESCRIPTION
**Pull request by olafurpg**
DenotationOps is 99% copy-pasta from scalahost.

Review @jvican.

### CODE REVIEW

File: sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala [hunk 1]
```diff
@@ -40,30 +40,32 @@ trait SbthostPipeline extends DatabaseOps { self: SbthostPlugin =>
...
+        val denots = ListBuffer.newBuilder[s.SymbolDenotation]
+        def getNames(): Unit = {
+          object traverser extends g.Traverser {
+            def isValidSymbol(symbol: g.Symbol) =
+              symbol.ne(null) && symbol != g.NoSymbol
+            override def traverse(tree: g.Tree): Unit = {
+              if (tree.pos.isDefined &&
+                  tree.hasSymbol &&
+                  isValidSymbol(tree.symbol) &&
+                  isValidSymbol(tree.symbol.owner)) {
```
Comment (jvican) [inline sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala:55]: What are we filtering here concretely? I guess it's syntactic members generated by scalac

File: sbthost/tests/src/test/scala/scala/meta/tests/SbthostTest.scala [hunk 2]
```diff
@@ -41,6 +41,30 @@ class SbthostTest extends org.scalatest.FunSuite {
...
+             |Denotations:
+             |_root_.java.lang.Object#`<init>`()V. => primaryctor <init>: ()Object
+             |_root_.sbthost. => package sbthost
+             |_root_.sbthost. => package sbthost
+             |_root_.sbthost.CompiledWithSbthost. => final object CompiledWithSbthost
+             |_root_.sbthost.CompiledWithSbthost. => final object CompiledWithSbthost
+             |_root_.sbthost.CompiledWithSbthost. => final object CompiledWithSbthost
+             |_root_.sbthost.CompiledWithSbthost. => final object CompiledWithSbthost
+             |_root_.sbthost.CompiledWithSbthost.`<init>`()V. => primaryctor <init>: ()sbthost.CompiledWithSbthost.type
+             |_root_.sbthost.CompiledWithSbthost.bar(I)I. => def bar: (x: Int)Int
```
Comment (jvican) [inline sbthost/tests/src/test/scala/scala/meta/tests/SbthostTest.scala]: 😂

**Review by jvican — APPROVED** [review]
LGTM

File: sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala [hunk 3]
```diff
@@ -43,7 +43,7 @@ trait SbthostPipeline extends DatabaseOps { self: SbthostPlugin =>
     override def newPhase(prev: Phase) = new StdPhase(prev) {
       def apply(unit: g.CompilationUnit): Unit = {
         val names = ListBuffer.newBuilder[s.ResolvedName]
-        val denots = ListBuffer.newBuilder[s.SymbolDenotation]
+        val denots = mutable.Map.empty[String, s.SymbolDenotation]
```
Comment (jvican) [inline sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala]: Have you considered using `perRunCaches`?

File: sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala (hunk 1 above)
Comment (olafurpg) [reply sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala:55]: There are many tree nodes without symbols, life in global can be rough.

File: sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala (hunk 3 above)
Comment (olafurpg) [reply sbthost/nsc/src/main/scala/scala/meta/internal/sbthost/SbthostPipeline.scala]: Nope, if performance ends up being a bottleneck then we can consider that.

**Pull request merged**